### DISCUSSION TITLE
S3 and volume replication

### DISCUSSION THREAD

**Discussion by candlerb** [seaweedfs/seaweedfs#1940] [Q&A] [answered]
I'm having difficultly understanding the comment [here](https://github.com/chrislusf/seaweedfs/wiki/Amazon-S3-API) and [here](https://github.com/chrislusf/seaweedfs/wiki/S3-Gateway-FAQ#can-not-upload-due-to-no-free-volumes-left):

-----

```
> fs.configure -locationPrefix=/buckets/ -volumeGrowthCount=1 -apply
```

This will add 1 physical volume when existing volumes are full. **If using replication, you will need to add more volumes.**

-----

It's the highlighted sentence I don't understand.

I thought that replication was already handled at the volume layer.  That is: a single volume, e.g. id "1234", can have replicas across multiple different physical servers.  It depends on the replication strategy selected when the volume is created.

However, the implication of the text above is that to get three-way replication in an S3 bucket I'd need at least `-volumeGrowthCount=3` - is that true ?!  If so, why?

I thought that the only reason for creating multiple volumes was to give a higher degree of concurrency for busy buckets.  That is, with `-volumeGrowthCount=1` the writes can be distributed 7 ways, and reads can be spread (7 x number of replicas) ways.  Is that not true?

**Comment by chrislusf** [comment, marked as answer]
"This will add 1 physical volume"

Seems you are confused about logical volume and physical volume. One logical volume can have multiple physical volumes (replicas).

**Reply by candlerb** [reply]
Then I don't understand why it's useful to specify physical volume count rather than logical volume count - since you can't make a logical volume without a full set of physical volumes.

I think I found the answer partly in `weed/shell/command_fs_configure.go`:

```
                        if *volumeGrowthCount%rp.GetCopyCount() != 0 {
                                return fmt.Errorf("volumeGrowthCount %d should be devided by replication copy count %d", *volumeGrowthCount, rp.GetCopyCount())
                        }
```

If I understand rightly, this means for example you can't set `-replication=001 -volumeGrowthCount=7`

But you could instead set `-replication=001 -volumeGrowthCount=6`, which means "when out of space, create three new volumes".  Is that correct?

**Reply by chrislusf** [reply]
> set `-replication=001 -volumeGrowthCount=6`, which means "when out of space, create three new volumes". Is that correct?

This is correct.

**Reply by candlerb** [reply]
Thank you. I've updated the wiki to clarify.
